File: cpp/c-api/src/core/vis.cpp
#include "dnv/vista/sdk/c/core/vis.h"

#include "../cast_internal.h"
#include "../error_internal.h"

using namespace dnv::vista::sdk;
using dnv::vista::sdk::c::fromGmodNode;
using dnv::vista::sdk::c::fromGmodPath;
using dnv::vista::sdk::c::fromLocalId;
using dnv::vista::sdk::c::fromLocalIdBuilder;
using dnv::vista::sdk::c::toGmodNode;
using dnv::vista::sdk::c::toGmodPath;
using dnv::vista::sdk::c::toLocalId;
using dnv::vista::sdk::c::toLocalIdBuilder;
using dnv::vista::sdk::c::toVis;

const dnv_vista_sdk_vis_t* dnv_vista_sdk_vis_instance(void)
{
    return reinterpret_cast<const dnv_vista_sdk_vis_t*>(&VIS::instance());
}

size_t dnv_vista_sdk_vis_version_count(const dnv_vista_sdk_vis_t* vis)
{
    if (vis == nullptr)
    {
        c::setLastErrorMessage("vis must not be null");
        return 0;
    }

    return toVis(vis)->versions().size();
}

const char* dnv_vista_sdk_vis_version_at(const dnv_vista_sdk_vis_t* vis, size_t index)
{
    if (vis == nullptr)
    {
        c::setLastErrorMessage("vis must not be null");
        return nullptr;
    }

    const auto versions = toVis(vis)->versions();
    if (index >= versions.size())
    {
        c::setLastErrorMessage("index out of range");
        return nullptr;
    }

    return VisVersions::toString(versions[index]).data();
}

const char* dnv_vista_sdk_vis_latest(const dnv_vista_sdk_vis_t* vis)
{
    if (vis == nullptr)
    {
        c::setLastErrorMessage("vis must not be null");
        return nullptr;
    }

    return VisVersions::toString(toVis(vis)->latest()).data();
}

const dnv_vista_sdk_gmod_t* dnv_vista_sdk_vis_gmod(const dnv_vista_sdk_vis_t* vis, const char* visVersion)
{
    if (vis == nullptr || visVersion == nullptr)
    {
        c::setLastErrorMessage("vis and visVersion must not be null");
        return nullptr;
    }

    const auto version = VisVersions::fromString(visVersion);
    if (!version.has_value())
    {
        c::setLastErrorMessage("unrecognized VIS version");
        return nullptr;
    }

    return reinterpret_cast<const dnv_vista_sdk_gmod_t*>(&toVis(vis)->gmod(*version));
}

const dnv_vista_sdk_codebooks_t* dnv_vista_sdk_vis_codebooks(const dnv_vista_sdk_vis_t* vis, const char* visVersion)
{
    if (vis == nullptr || visVersion == nullptr)
    {
        c::setLastErrorMessage("vis and visVersion must not be null");
        return nullptr;
    }

    const auto version = VisVersions::fromString(visVersion);
    if (!version.has_value())
    {
        c::setLastErrorMessage("unrecognized VIS version");
        return nullptr;
    }

    return reinterpret_cast<const dnv_vista_sdk_codebooks_t*>(&toVis(vis)->codebooks(*version));
}

const dnv_vista_sdk_locations_t* dnv_vista_sdk_vis_locations(const dnv_vista_sdk_vis_t* vis, const char* visVersion)
{
    if (vis == nullptr || visVersion == nullptr)
    {
        c::setLastErrorMessage("vis and visVersion must not be null");
        return nullptr;
    }

    const auto version = VisVersions::fromString(visVersion);
    if (!version.has_value())
    {
        c::setLastErrorMessage("unrecognized VIS version");
        return nullptr;
    }

    return reinterpret_cast<const dnv_vista_sdk_locations_t*>(&toVis(vis)->locations(*version));
}

dnv_vista_sdk_gmod_node_t* dnv_vista_sdk_vis_convert_node(
    const dnv_vista_sdk_vis_t* vis,
    const char* sourceVersion,
    const dnv_vista_sdk_gmod_node_t* sourceNode,
    const char* targetVersion)
{
    if (vis == nullptr || sourceVersion == nullptr || sourceNode == nullptr || targetVersion == nullptr)
    {
        c::setLastErrorMessage("vis, sourceVersion, sourceNode and targetVersion must not be null");
        return nullptr;
    }

    const auto source = VisVersions::fromString(sourceVersion);
    const auto target = VisVersions::fromString(targetVersion);
    if (!source.has_value() || !target.has_value())
    {
        c::setLastErrorMessage("unrecognized VIS version");
        return nullptr;
    }

    auto node = toVis(vis)->convertNode(*source, *toGmodNode(sourceNode), *target);
    if (!node.has_value())
    {
        c::setLastErrorMessage("conversion failed");
    }

    return fromGmodNode(std::move(node));
}

dnv_vista_sdk_gmod_path_t* dnv_vista_sdk_vis_convert_path(
    const dnv_vista_sdk_vis_t* vis,
    const char* sourceVersion,
    const dnv_vista_sdk_gmod_path_t* sourcePath,
    const char* targetVersion)
{
    if (vis == nullptr || sourceVersion == nullptr || sourcePath == nullptr || targetVersion == nullptr)
    {
        c::setLastErrorMessage("vis, sourceVersion, sourcePath and targetVersion must not be null");
        return nullptr;
    }

    const auto source = VisVersions::fromString(sourceVersion);
    const auto target = VisVersions::fromString(targetVersion);
    if (!source.has_value() || !target.has_value())
    {
        c::setLastErrorMessage("unrecognized VIS version");
        return nullptr;
    }

    auto path = toVis(vis)->convertPath(*source, *toGmodPath(sourcePath), *target);
    if (!path.has_value())
    {
        c::setLastErrorMessage("conversion failed");
    }

    return fromGmodPath(std::move(path));
}

dnv_vista_sdk_local_id_builder_t* dnv_vista_sdk_vis_convert_local_id_builder(
    const dnv_vista_sdk_vis_t* vis, const dnv_vista_sdk_local_id_builder_t* sourceLocalId, const char* targetVersion)
{
    if (vis == nullptr || sourceLocalId == nullptr || targetVersion == nullptr)
    {
        c::setLastErrorMessage("vis, sourceLocalId and targetVersion must not be null");
        return nullptr;
    }

    const auto target = VisVersions::fromString(targetVersion);
    if (!target.has_value())
    {
        c::setLastErrorMessage("unrecognized VIS version");
        return nullptr;
    }

    try
    {
        auto localId = toVis(vis)->convertLocalId(*toLocalIdBuilder(sourceLocalId), *target);
        if (!localId.has_value())
        {
            c::setLastErrorMessage("conversion failed");
            return nullptr;
        }

        return fromLocalIdBuilder(std::move(*localId));
    }
    catch (const std::exception& e)
    {
        c::setLastErrorMessage(e.what());
        return nullptr;
    }
}

dnv_vista_sdk_local_id_t* dnv_vista_sdk_vis_convert_local_id(
    const dnv_vista_sdk_vis_t* vis, const dnv_vista_sdk_local_id_t* sourceLocalId, const char* targetVersion)
{
    if (vis == nullptr || sourceLocalId == nullptr || targetVersion == nullptr)
    {
        c::setLastErrorMessage("vis, sourceLocalId and targetVersion must not be null");
        return nullptr;
    }

    const auto target = VisVersions::fromString(targetVersion);
    if (!target.has_value())
    {
        c::setLastErrorMessage("unrecognized VIS version");
        return nullptr;
    }

    try
    {
        auto localId = toVis(vis)->convertLocalId(*toLocalId(sourceLocalId), *target);
        if (!localId.has_value())
        {
            c::setLastErrorMessage("conversion failed");
            return nullptr;
        }

        return fromLocalId(std::move(localId));
    }
    catch (const std::exception& e)
    {
        c::setLastErrorMessage(e.what());
        return nullptr;
    }
}

File: bindings/rust/vista-sdk/src/core/vis.rs
use std::ffi::{CStr, CString};
use std::str::FromStr;

use crate::core::codebooks::Codebooks;
use crate::core::error::{last_error, VistaError};
use crate::core::gmod::Gmod;
use crate::core::gmod_node::{GmodNode, OwnedGmodNode};
use crate::core::gmod_path::{GmodPath, OwnedGmodPath};
use crate::core::local_id::{LocalId, OwnedLocalId};
use crate::core::local_id_builder::{LocalIdBuilder, OwnedLocalIdBuilder};
use crate::core::locations::Locations;
use crate::core::vis_version::VisVersion;
use crate::ffi::core::vis as ffi;

/// Central entry point for the Vista SDK.
///
/// Wraps a borrowed pointer to the C++ singleton, valid for the lifetime of the program.
pub struct Vis(*const ffi::dnv_vista_sdk_vis_t);

unsafe impl Send for Vis {}
unsafe impl Sync for Vis {}

impl Vis {
    /// Returns the VIS singleton instance.
    pub fn instance() -> Vis {
        Vis(unsafe { ffi::dnv_vista_sdk_vis_instance() })
    }

    /// Returns all available VIS versions, in ascending order.
    pub fn versions(&self) -> Vec<VisVersion> {
        let count = unsafe { ffi::dnv_vista_sdk_vis_version_count(self.0) };
        (0..count)
            .filter_map(|i| {
                let ptr = unsafe { ffi::dnv_vista_sdk_vis_version_at(self.0, i) };
                if ptr.is_null() {
                    return None;
                }
                let s = unsafe { CStr::from_ptr(ptr) }.to_str().ok()?;
                VisVersion::from_str(s).ok()
            })
            .collect()
    }

    /// Returns the latest VIS version.
    pub fn latest(&self) -> VisVersion {
        let ptr = unsafe { ffi::dnv_vista_sdk_vis_latest(self.0) };
        assert!(!ptr.is_null(), "dnv_vista_sdk_vis_latest returned NULL");
        let s = unsafe { CStr::from_ptr(ptr) }
            .to_str()
            .expect("invalid UTF-8 in latest version");
        VisVersion::from_str(s).expect("unrecognized latest version string")
    }

    /// Returns the codebooks for `version`.
    pub fn codebooks(&self, version: VisVersion) -> Result<&Codebooks, VistaError> {
        let s = CString::new(version.as_str()).expect("invalid UTF-8 in VisVersion");
        let ptr = unsafe { ffi::dnv_vista_sdk_vis_codebooks(self.0, s.as_ptr()) };
        if ptr.is_null() {
            Err(last_error())
        } else {
            Ok(Codebooks::from_ptr(ptr))
        }
    }

    /// Returns the locations for `version`.
    pub fn locations(&self, version: VisVersion) -> Result<&Locations, VistaError> {
        let s = CString::new(version.as_str()).expect("invalid UTF-8 in VisVersion");
        let ptr = unsafe { ffi::dnv_vista_sdk_vis_locations(self.0, s.as_ptr()) };
        if ptr.is_null() {
            Err(last_error())
        } else {
            Ok(Locations::from_ptr(ptr))
        }
    }

    /// Returns the Gmod for `version`.
    pub fn gmod(&self, version: VisVersion) -> Result<&Gmod, VistaError> {
        let s = CString::new(version.as_str()).expect("invalid UTF-8 in VisVersion");
        let ptr = unsafe { ffi::dnv_vista_sdk_vis_gmod(self.0, s.as_ptr()) };
        if ptr.is_null() {
            Err(last_error())
        } else {
            Ok(Gmod::from_ptr(ptr))
        }
    }

    /// Converts a Gmod node from `source_version` to `target_version`.
    pub fn convert_node(
        &self,
        source_version: VisVersion,
        source_node: &GmodNode,
        target_version: VisVersion,
    ) -> Result<OwnedGmodNode, VistaError> {
        let source_version_c =
            CString::new(source_version.as_str()).expect("invalid UTF-8 in VisVersion");
        let target_version_c =
            CString::new(target_version.as_str()).expect("invalid UTF-8 in VisVersion");
        let ptr = unsafe {
            ffi::dnv_vista_sdk_vis_convert_node(
                self.0,
                source_version_c.as_ptr(),
                source_node.as_ffi_ptr(),
                target_version_c.as_ptr(),
            )
        };
        if ptr.is_null() {
            Err(last_error())
        } else {
            Ok(OwnedGmodNode::with_drop(
                ptr,
                crate::ffi::core::gmod_node::dnv_vista_sdk_gmod_node_free,
            ))
        }
    }

    /// Converts a Gmod path from `source_version` to `target_version`.
    pub fn convert_path(
        &self,
        source_version: VisVersion,
        source_path: &GmodPath,
        target_version: VisVersion,
    ) -> Result<OwnedGmodPath, VistaError> {
        let source_version_c =
            CString::new(source_version.as_str()).expect("invalid UTF-8 in VisVersion");
        let target_version_c =
            CString::new(target_version.as_str()).expect("invalid UTF-8 in VisVersion");
        let ptr = unsafe {
            ffi::dnv_vista_sdk_vis_convert_path(
                self.0,
                source_version_c.as_ptr(),
                source_path.as_ffi_ptr(),
                target_version_c.as_ptr(),
            )
        };
        if ptr.is_null() {
            Err(last_error())
        } else {
            Ok(OwnedGmodPath::from_owned_ptr(ptr))
        }
    }

    /// Converts a LocalIdBuilder to `target_version`.
    pub fn convert_local_id_builder(
        &self,
        source_local_id: &LocalIdBuilder,
        target_version: VisVersion,
    ) -> Result<OwnedLocalIdBuilder, VistaError> {
        let target_version_c =
            CString::new(target_version.as_str()).expect("invalid UTF-8 in VisVersion");
        let ptr = unsafe {
            ffi::dnv_vista_sdk_vis_convert_local_id_builder(
                self.0,
                source_local_id.as_ffi_ptr(),
                target_version_c.as_ptr(),
            )
        };
        if ptr.is_null() {
            Err(last_error())
        } else {
            Ok(OwnedLocalIdBuilder(ptr))
        }
    }

    /// Converts a LocalId to `target_version`.
    pub fn convert_local_id(
        &self,
        source_local_id: &LocalId,
        target_version: VisVersion,
    ) -> Result<OwnedLocalId, VistaError> {
        let target_version_c =
            CString::new(target_version.as_str()).expect("invalid UTF-8 in VisVersion");
        let ptr = unsafe {
            ffi::dnv_vista_sdk_vis_convert_local_id(
                self.0,
                source_local_id.as_ffi_ptr(),
                target_version_c.as_ptr(),
            )
        };
        if ptr.is_null() {
            Err(last_error())
        } else {
            Ok(OwnedLocalId::from_owned_ptr(ptr))
        }
    }
}
